src/control/tools/CircleHandler.cpp

```diff
@@ -18,53 +19,90 @@ CircleHandler::~CircleHandler()
 	XOJ_RELEASE_TYPE(CircleHandler);
 }
 
-void CircleHandler::drawShape(Point& c, bool shiftDown)
+
+void CircleHandler::drawShape(Point& c, const PositionInputData& pos)
 {
 	XOJ_CHECK_TYPE(CircleHandler);
-
+	
+	this->currPoint = c;
+		
 	/**
-	 * Snap first point to grid (if enabled)
+	 * Snap point to grid (if enabled - Alt key pressed will toggle)
 	 */
-	if (!shiftDown && xournal->getControl()->getSettings()->isSnapGrid())
+	Settings* settings = xournal->getControl()->getSettings();
+	if ( pos.isAltDown() != settings->isSnapGrid())
 	{
-		Point firstPoint = stroke->getPoint(0);
-		snapToGrid(firstPoint.x,firstPoint.y);
-		stroke->setFirstPoint(firstPoint.x,firstPoint.y);
+		snapToGrid(c.x,c.y);
 	}
 
-	int count = stroke->getPointCount();
-
-	if (count < 2)
+	if (!this->started) //initialize first point
 	{
-		stroke->addPoint(c);
+		this->startPoint = c;
+		this->started = true;
+		stroke->addPoint(c);	//avoid complaints about <2 points.
 	}
 	else
-	{
-		Point p = stroke->getPoint(0);
-		if (xournal->getControl()->getSettings()->isSnapGrid())
+	{		
+		double width = c.x - this->startPoint.x;
+		double height = c.y - this->startPoint.y;
+		
+		
+		
+		this->modShift = pos.isShiftDown();
+		this->modControl = pos.isControlDown();
+		
+		if ( settings->getDrawDirModsEnabled()) //change modifiers based on draw dir
 		{
-			snapToGrid(c.x,c.y);
+			this->modifyModifiersByDrawDir(width, height, true);
+		}
+	
+		if (this->modShift )	// make circle
+		{
+			int signW = width>0?1:-1;
+			int signH = height>0?1:-1;
+			width = MAX( width*signW, height*signH) * signW;	
+			height = (width * signW) * signH;
 		}
-		// set resolution proportional to radius
-		double diameter = sqrt(pow(c.x - p.x, 2.0) + pow(c.y - p.y, 2.0));
-		int npts = (int) (diameter * 2.0);
-		double center_x = (c.x + p.x) / 2.0;
-		double center_y = (c.y + p.y) / 2.0;
-		double angle = atan2((c.y - p.y), (c.x - p.x));
 
+		double diameterX, diameterY;
```

Comment:
C++ does not initialize vars, and C++ not always warns about unitialized vars (like Java does) therefore I would initialize all vars every time.

The optimizer will remove unused code, therefore it's not a performance issue etc.